golem-worker-service-base/src/http/router.rs

```diff
@@ -0,0 +1,471 @@
+                        children: new_child_children,
+                        data: new_child_data,
+                    };
+
+                    child
+                        .children
+                        .insert(new_child.pattern[0].clone(), new_child);
+                    child.push(&path[common_prefix_len..], data);
+                }
+                return;
+            }
+        }
+
+        let new_node = RadixNode::new(path.to_vec());
+        self.children.insert(path[0].clone(), new_node);
+        if let Some(child) = self.children.get_mut(&path[0]) {
+            child.data = Some(data);
+        }
+    }
+
+    pub fn get(&self, path: &[Pattern]) -> Option<&T> {
+        if path.is_empty() {
+            return self.data.as_ref();
+        }
+
+        if let Some(child) = self.children.get(&path[0]) {
+            if path.starts_with(&child.pattern) {
+                return child.get(&path[child.pattern.len()..]);
+            }
+        }
+
+        None
+    }
+
+    pub fn matches<'a, 'b>(&'a self, path: &'b str) -> Option<MatchResult<'a, 'b, T>> {
+        let path = path
+            .trim()
+            .trim_start_matches('/')
+            .trim_end_matches('/')
+            .split('/')
+            .collect::<Vec<_>>();
+
+        let mut node = self;
+        let mut path_segments = path.as_slice();
+        let mut variables = Vec::new();
+
+        loop {
+            if path_segments.is_empty() {
+                let new_values = extract_values(path_segments, &node.pattern);
+                variables.extend(new_values);
+                return node
+                    .data
+                    .as_ref()
+                    .map(|data| MatchResult { data, variables });
+            }
+
+            let current_segment = path_segments[0];
+
+            let next_child = node
+                .children
+                .iter()
+                .find(|(pattern, _)| match pattern {
+                    Pattern::Static(s) => s == current_segment,
+                    Pattern::Variable => true,
+                })
+                .map(|(_, child)| child);
+
+            if let Some(child) = next_child {
+                let common_prefix_len = child
+                    .pattern
+                    .iter()
+                    .zip(path_segments.iter())
+                    .take_while(|(a, b)| match a {
+                        Pattern::Static(s) => s == *b,
+                        Pattern::Variable => true,
+                    })
+                    .count();
+
+                if common_prefix_len == child.pattern.len() {
+                    let path_start = path.len() - path_segments.len();
+                    let path_end = path_start + common_prefix_len;
+
+                    let new_variables = extract_values(&path[path_start..path_end], &child.pattern);
+                    variables.extend(new_variables);
+
+                    path_segments = &path_segments[common_prefix_len..];
```

Comment:
I would use `.get`, or will that index out never happen?

Reply:
We will probably use `.matches` in the route resolution. because that will extract the variables from path